### PR TITLE
Feature/ND/213/HAL-I2C-API 

### PR DESCRIPTION
**Pull request by colton-smith**
Created a wrapper for the HAL Generic I2C library functionality. Due to the simplicity of the class, there is no manager class, only an interface class. I2CInterface.h and I2CInterface.cpp are located under Common/PeripheralInterfaces/include and src respectively.

The primary HAL-I2C functionality encapsulated is:
- Send an amount of data in blocking mode using masterRecieve and masterTransmit

- Transmit/Receive to and from internal device registers using memWrite and memReceive

The HAL Generic I2C documentation can be found on page 441 here: 
https://www.st.com/content/ccc/resource/technical/document/user_manual/2f/71/ba/b8/75/54/47/cf/DM00105879.pdf/files/DM00105879.pdf/jcr:content/translations/en.DM00105879.pdf

### CODE REVIEW

**Comment by colton-smith** [comment]
I was wondering what the includes should be for the I2CInterface.h file should be, probably should include the HAL I2C library header file, but im not sure.

**Comment by andrewnash** [comment]
Can't comment on functionality, but style-wise looks clean. Good use of comments and good commit messages 🙂

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#ifndef H_I2CINTERFACE
+#define H_I2CINTERFACE
```
Comment (ShaneTWilliams) [inline OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: Just to keep it consistent we can use FILENAME_H. Try I2CINTERFACE_H

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#include "main.h"
+
+class I2CInterface
+{
+    private:
+
+    // Pointer to I2C Handle Type Def structure contains the config info for the specified I2C.
+    I2C_HandleTypeDef *m_hi2c;
+    uint8_t m_dataRecieved[8]; 
+    uint8_t m_dataBufferLength = 8;
```
Comment (ShaneTWilliams) [inline OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: Should we do something about this 8? It kinda seems like a magic number but idk how else to do it

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    uint32_t m_timeout;
```
Comment (ShaneTWilliams) [inline OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: This value doesn't get set and is only used in the `isDeviceReady` method. I'd say your best bet is to have `isDeviceReady ` take timeout as a param like the other methods do

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    
+    // Returns a pointer to the data recieved buffer
+    uint8_t* getRecievedData() {  return m_dataRecieved; }
+
+    // Get state, and error methods
+    HAL_I2C_StateTypeDef getState() { return HAL_I2C_GetState(m_hi2c); }
+    uint32_t getError() { return HAL_I2C_GetError(m_hi2c); }
+
+    // Getter and setter for I2C addressing mode
+    void setAddressingMode(uint32_t mode) { (*m_hi2c).Init.AddressingMode = mode; }
```
Comment (ShaneTWilliams) [inline OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h:28]: @stuckatmarine how do you feel about defining one-liners in the header? Maybe only getters and class attribute setters?

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    uint32_t getAddressingMode() { return (*m_hi2c).Init.AddressingMode; }
+
+    // Master transmit and recieve functions (blocking mode)
+    HAL_StatusTypeDef masterTransmit(uint16_t address, uint8_t* data, uint32_t timeout, uint8_t size);   
+    HAL_StatusTypeDef masterRecieve(uint16_t address, uint32_t timeout);
```
Comment (ShaneTWilliams) [inline OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h:33]: Look into a non-blocking send function as waiting on peripherals could slow down the loop

File: OnPod/Node/Common/PeripheralInterfaces/src/I2CInterface.cpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
+#include "I2CInterface.h"
+
+I2CInterface::I2CInterface(I2C_HandleTypeDef *hi2c)
+{
+    m_hi2c = hi2c;
```
Comment (ShaneTWilliams) [inline OnPod/Node/Common/PeripheralInterfaces/src/I2CInterface.cpp]: Use an initializer list for this

Comment (ShaneTWilliams) [reply OnPod/Node/Common/PeripheralInterfaces/src/I2CInterface.cpp]: we get it, you write python

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    // Reading / writing from device internal memory addresses 
+    HAL_StatusTypeDef memWrite(uint8_t* data, uint16_t devAddress, uint16_t memAddress, uint16_t memAddressSize, uint8_t size, uint32_t timeout);
```
Comment (ShaneTWilliams) [inline OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: Watch the line length here, it's typical to start wrapping lines if you go over 80 characters

**Review by ShaneTWilliams — CHANGES_REQUESTED** [review]
Looks sick! Besides the non-blocking receive function everything is pretty much style

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h (hunk 3 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: I only meant to highlight line 14, this is about `m_timeout`

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h (hunk 5 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h:33]: Never mind this, HAL's "blocking" mode is actually a polling mode which is dope

**Comment by ShaneTWilliams** [comment]
> I was wondering what the includes should be for the I2CInterface.h file should be, probably should include the HAL I2C library header file, but im not sure.

If it compiles without including this then that means that PlatformIO is including it for you. Should be good to leave it out

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h (hunk 2 above)
Comment (colton-smith) [reply OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: Other than std::vector or dynamically allocating the memory I can't see a way to dynamically size the buffer.

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h (hunk 7 above)
Comment (colton-smith) [reply OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: Wrapped really long lines to better adhere to the 80 char limit. If readability was severy reduced when wrapped and the length was 80-100, left untouched.

File: OnPod/Node/Common/PeripheralInterfaces/src/I2CInterface.cpp (hunk 6 above)
Comment (colton-smith) [reply OnPod/Node/Common/PeripheralInterfaces/src/I2CInterface.cpp]: self.status = done

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h (hunk 1 above)
Comment (colton-smith) [reply OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: Went with discussed NODE_I2CINTERFACE_H

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h (hunk 2 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h]: Should be good for now

File: OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h (hunk 4 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/PeripheralInterfaces/include/I2CInterface.h:28]: This should be fine, you can keep it

**Review by ShaneTWilliams — APPROVED** [review]
Everything looks good to merge

**Pull request merged**